xapian-core/diversify/diversify.cc

```diff
@@ -0,0 +1,195 @@
+/** @file diversify.cc
+ *  @brief Diversification API
+ */
+/* Copyright (C) 2018 Uppinder Chugh
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+
+#include "xapian/diversify.h"
+
+#include "xapian/error.h"
+#include "debuglog.h"
+
+#include <vector>
+#include <limits>
+#include <algorithm>
+#include <cmath>
+
+using namespace Xapian;
+using namespace std;
+
+Diversify::Diversify(unsigned int k_,
+		     double lambda_,
+		     double b_,
+		     double sigma_sqr_)
+    : k(k_), lambda(lambda_), b(b_), sigma_sqr(sigma_sqr_)
+{
+    LOGCALL_CTOR(API, "Diversify", k_ | lambda_ | b_ | sigma_sqr_);
+}
+
+string
+Diversify::get_description() const
+{
+    return "Diversify()";
+}
+
+void
+Diversify::initialise_points(const MSet &source)
+{
+    LOGCALL_VOID(API, "Diversify::initialise_points", source);
+    TermListGroup tlg(source);
+    for (MSetIterator it = source.begin(); it != source.end(); ++it) {
+	points.push_back(Point(tlg, it.get_document()));
+	weights[it.get_document().get_docid()] = it.get_weight();
+    }
+}
+
+void
+Diversify::compute_similarities()
+{
+    LOGCALL_VOID(API, "Diversify::compute_similarities", NO_ARGS);
+    Xapian::CosineDistance d;
+    for (unsigned int i = 0; i < points.size() - 1; ++i) {
+	for (unsigned int j = i + 1; j < points.size(); ++j) {
+	    Xapian::docid docid_i = points[i].get_document().get_docid();
+	    Xapian::docid docid_j = points[j].get_document().get_docid();
+	    auto key_ij = make_pair(docid_i, docid_j);
+	    auto key_ji = make_pair(docid_j, docid_i);
+	    double sim = d.similarity(points[i], points[j]);
+	    pairwise_sim[key_ij] = pairwise_sim[key_ji] = sim;
```

Comment:
If you only stored the similarly for `(x, y)` where `x` <= `y` then you would roughly halve the amount of storage needed for `pairwise_sim` and greatly reduce the cache pressure when you doing all the lookups in it.  The only cost would be that you need to compare the two docids when doing a lookup, and do the lookup such that the smaller one is first in the pair.  I'd expect that's a significant win overall.

Reply:
Yes, it definitely is. Thanks for pointing this out, pushed required changes in new commit.